clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1036,6 +1010,55 @@ void Sema::MarkDevice(void) {
     }
   }
 }
+//
+// Do we know that we will eventually codegen the given function?
+static bool isKnownEmitted(Sema &S, FunctionDecl *FD) {
+  if (!FD)
+    return true; // Seen in LIT testing
+
+  if (FD->hasAttr<SYCLDeviceAttr>() ||
+      FD->hasAttr<SYCLKernelAttr>())
+     return true;
+
+  // Templates are emitted when they're instantiated.
+  if (FD->isDependentContext())
+    return false;
+
+  // Otherwise, the function is known-emitted if it's in our set of
+  // known-emitted functions.
+  return S.DeviceKnownEmittedFns.count(FD) > 0;
+}
+
+Sema::DeviceDiagBuilder Sema::SYCLDiagIfDeviceCode(SourceLocation Loc,
+                                                   unsigned DiagID) {
+  assert(getLangOpts().SYCLIsDevice &&
+    "Should only be called during SYCL compilation");
+  DeviceDiagBuilder::Kind DiagKind = [this] {
+    if (isKnownEmitted(*this, dyn_cast<FunctionDecl>(CurContext)))
+      return DeviceDiagBuilder::K_ImmediateWithCallStack;
+    else
```

Comment:
No `else` after `return`, please.
http://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

Reply:
Thank you for catching that.   Fixed.